engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/hv/UUIDValidator.java

```diff
@@ -0,0 +1,174 @@
+
+/**
+ * Checks that the annotated character sequence is a valid
+ * <a href="https://en.wikipedia.org/wiki/Universally_unique_identifier">UUID</a>.
+ *
+ * @author Daniel Heid
+ */
+public class UUIDValidator implements ConstraintValidator<UUID, CharSequence> {
+
+	private static final int[] GROUP_LENGTHS = { 8, 4, 4, 4, 12 };
+
+	private boolean allowEmpty;
+
+	private boolean allowNil;
+
+	private int[] version;
+
+	private int[] variant;
+
+	private LetterCase letterCase;
+
+	@Override
+	public void initialize(UUID constraintAnnotation) {
+		allowEmpty = constraintAnnotation.allowEmpty();
+		allowNil = constraintAnnotation.allowNil();
+		version = checkAndSortMultiOptionParameter( constraintAnnotation.version(), "version", 1, 15 );
+		variant = checkAndSortMultiOptionParameter( constraintAnnotation.variant(), "variant", 0, 2 );
+		letterCase = constraintAnnotation.letterCase();
+		Contracts.assertNotNull( letterCase, MESSAGES.parameterMustNotBeNull( "letterCase" ) );
+	}
+
+	private static int[] checkAndSortMultiOptionParameter(
+			int[] values,
+			String parameterName,
+			int minimum,
+			int maximum) {
+		Contracts.assertNotNull( values, MESSAGES.parameterMustNotBeNull( parameterName ) );
+		Contracts.assertNotEmpty( values, MESSAGES.parameterMustNotBeEmpty( parameterName ) );
+		for ( int value : values ) {
+			Contracts.assertTrue(
+					value >= minimum,
+					MESSAGES.parameterShouldBeGreaterThanOrEqualTo( parameterName, minimum )
+			);
+			Contracts.assertTrue(
+					value <= maximum,
+					MESSAGES.parameterShouldBeLessThanOrEqualTo( parameterName, maximum )
+			);
+		}
+		Arrays.sort( values );
+		return values;
+	}
+
+	@Override
+	public boolean isValid(CharSequence value, ConstraintValidatorContext context) {
+		if ( value == null ) {
+			return true;
+		}
+		int valueLength = value.length();
+		if ( valueLength == 0 ) {
+			return allowEmpty;
+		}
+		else if ( valueLength != 36 ) {
+			return false;
+		}
+
+		int groupIndex = 0;
+		int groupLength = 0;
+		int checksum = 0;
+		int version = -1;
+		int variant = -1;
+		for ( int charIndex = 0; charIndex < valueLength; charIndex++ ) {
+
+			char ch = value.charAt( charIndex );
+
+			if ( ch == '-' ) {
+				groupIndex++;
+				groupLength = 0;
+			}
+			else {
+
+				groupLength++;
+				if ( groupLength > GROUP_LENGTHS[groupIndex] ) {
+					return false;
+				}
+
+				int numericValue = Character.digit( ch, 16 );
+				if ( numericValue == -1 ) {
+					// not a hex digit
+					return false;
+				}
+				if ( letterCase == LetterCase.LOWER_CASE && numericValue > 9 && !Character.isLowerCase( ch ) ) {
+					return false;
+				}
+				if ( letterCase == LetterCase.UPPER_CASE && numericValue > 9 && !Character.isUpperCase( ch ) ) {
+					return false;
+				}
+				checksum += numericValue;
+				version = extractVersion( version, charIndex, numericValue );
+				variant = extractVariant( variant, charIndex, numericValue );
+
+			}
+
+		}
```

Comment:
This is not a request for a change, but I just wanted to share how I would have expected this to look like in case you want to change the implementation.
I would have expected a constant loop from 0 to 35 with a big switch covering all indices and asserting a valid char is used. Something like this:

```suggestion
		LOOP: for ( int charIndex = 0; charIndex < 36; charIndex++ ) {

			char ch = value.charAt( charIndex );

			switch ( charIndex ) {
				// Handle hyphens
				case 9:
				case 14:
				case 19:
				case 24:
					if ( ch != '-' ) {
						return false;
					}
					continue LOOP;
				// Handle M and N parts
				case 15:
					version = extractVersion( ch );
					break;
				case 20:
					variant = extractVariant( ch );
					break;
			}

			int numericValue = Character.digit( ch, 16 );
			if ( numericValue == -1 ) {
				// not a hex digit
				return false;
			}
			switch ( letterCase ) {
				case LOWER_CASE:
					if ( numericValue > 9 && !Character.isLowerCase( ch ) ) {
						return false;
					}
					break;
				case UPPER_CASE :
					if ( numericValue > 9 && !Character.isUpperCase( ch ) ) {
						return false;
					}
					break;
			}
			checksum += numericValue;
		}
```

Reply:
Since the chars have to be in ASCII encoding, you could go even a step further if you like by introducing an array for the character code points.

```java
private static final byte[] DIGITS = new byte[] {
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, -1, -1,
        -1, -1, -1, -1, -1, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 10, 11, 12,
        13, 14, 15
};
int charAsInt = ch;
int numericValue = charAsInt >= 0 && charAsInt < DIGITS.length ? DIGITS[charAsInt] : -1;
if ( numericValue == -1 ) {
	// not a hex digit
	return false;
}

switch ( letterCase ) {
	case LOWER_CASE:
		if ( ch > 'A' && ch < 'F' ) {
			return false;
		}
		break;
	case UPPER_CASE :
		if ( ch > 'a' && ch < 'f' ) {
			return false;
		}
		break;
}
```

Reply:
Tbh Christian, I find Daniel's approach more concise and I don't see how your suggestion actually improves upon it?

Reply:
I like this creative solution! But if you don't mind I'll stick to my solution, because it's easier to understand from my point of view.